File: packages/snow_draw_arrow_core/lib/src/arrow_relation_core.dart
class BoundRelationEntry {
  const BoundRelationEntry({required this.id, required this.type});

  final String id;
  final String type;
}

List<TEntry>? _normalizeRelationEntries<TEntry extends BoundRelationEntry>(
  List<TEntry>? entries,
) => entries;

bool areBoundRelationEntriesEqual<TEntry extends BoundRelationEntry>(
  List<TEntry>? left,
  List<TEntry>? right,
) {
  final normalizedLeft = _normalizeRelationEntries(left);
  final normalizedRight = _normalizeRelationEntries(right);

  if (identical(normalizedLeft, normalizedRight)) {
    return true;
  }
  if (normalizedLeft == null || normalizedRight == null) {
    return false;
  }
  if (normalizedLeft.length != normalizedRight.length) {
    return false;
  }

  for (var index = 0; index < normalizedLeft.length; index += 1) {
    final leftEntry = normalizedLeft[index];
    final rightEntry = normalizedRight[index];
    if (leftEntry.id != rightEntry.id || leftEntry.type != rightEntry.type) {
      return false;
    }
  }

  return true;
}

List<TEntry>? mergeBoundRelationEntries<TEntry extends BoundRelationEntry>({
  required List<TEntry>? entries,
  required String targetType,
  required List<String> targetIds,
}) {
  final normalized = entries ?? <TEntry>[];
  final preserved = normalized
      .where((entry) => entry.type != targetType)
      .toList();
  final next = <TEntry>[
    ...preserved,
    ...targetIds.map(
      (id) => BoundRelationEntry(id: id, type: targetType) as TEntry,
    ),
  ];

  return next.isNotEmpty ? next : null;
}

List<TEntry>?
mergeArrowBoundRelationEntries<TEntry extends BoundRelationEntry>({
  required List<TEntry>? entries,
  required List<String> boundArrowIds,
}) => mergeBoundRelationEntries<TEntry>(
  entries: entries,
  targetType: 'arrow',
  targetIds: boundArrowIds,
);

File: packages/snow_draw_rust_engine/src/draw/elements/types/arrow/core/arrow_relation_core.rs
#![allow(dead_code)]

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct BoundRelationEntry {
    pub id: String,
    pub entry_type: String,
}

pub fn are_bound_relation_entries_equal(
    left: Option<&[BoundRelationEntry]>,
    right: Option<&[BoundRelationEntry]>,
) -> bool {
    match (left, right) {
        (None, None) => true,
        (Some(left), Some(right)) if left.len() == right.len() => {
            left.iter().zip(right).all(|(a, b)| a == b)
        }
        _ => false,
    }
}

pub fn merge_bound_relation_entries(
    entries: Option<&[BoundRelationEntry]>,
    target_type: &str,
    target_ids: &[String],
) -> Option<Vec<BoundRelationEntry>> {
    let mut merged = entries
        .unwrap_or_default()
        .iter()
        .filter(|entry| entry.entry_type != target_type)
        .cloned()
        .collect::<Vec<_>>();
    merged.extend(target_ids.iter().cloned().map(|id| BoundRelationEntry {
        id,
        entry_type: target_type.to_string(),
    }));
    (!merged.is_empty()).then_some(merged)
}

pub fn merge_arrow_bound_relation_entries(
    entries: Option<&[BoundRelationEntry]>,
    bound_arrow_ids: &[String],
) -> Option<Vec<BoundRelationEntry>> {
    merge_bound_relation_entries(entries, "arrow", bound_arrow_ids)
}
